File: packages/runtime/ios/src/LynxCapacitorBridge.m
#import "LynxCapacitorBridge.h"
#import <Lynx/LynxContext.h>
#import <LynxCapacitorRuntime/LynxCapacitorRuntime-Swift.h>

@interface LynxCapacitorBridge ()
@property(nonatomic, strong) LynxCapacitorRuntime *runtime;
@property(nonatomic, weak) LynxContext *lynxContext;
@end

@implementation LynxCapacitorBridge

+ (NSString *)name {
  return @"CapacitorBridge";
}

+ (NSDictionary<NSString *, NSString *> *)methodLookup {
  return @{
    @"handleCall" : NSStringFromSelector(@selector(handleCall:callback:)),
    @"getPluginHeaders" : NSStringFromSelector(@selector(getPluginHeaders)),
    @"getPlatform" : NSStringFromSelector(@selector(getPlatform)),
  };
}

- (instancetype)init {
  if (self = [super init]) {
    if ([NSThread isMainThread]) {
      _runtime = [[LynxCapacitorRuntime alloc] init];
    } else {
      dispatch_sync(dispatch_get_main_queue(), ^{
        self->_runtime = [[LynxCapacitorRuntime alloc] init];
      });
    }
  }
  return self;
}

- (instancetype)initWithLynxContext:(LynxContext *)context {
  if (self = [self init]) {
    self.lynxContext = context;
    __weak typeof(self) weakSelf = self;
    [self.runtime setResultHandler:^(NSString *resultJson) {
      LynxContext *strongContext = weakSelf.lynxContext;
      if (strongContext == nil) {
        return;
      }
      [strongContext sendGlobalEvent:@"lynx-capacitor-result" withParams:@[ resultJson ]];
    }];
  }
  return self;
}

- (NSString *)getPlatform {
  return [self.runtime getPlatform];
}

- (NSString *)getPluginHeaders {
  return [self.runtime getPluginHeaders];
}

- (void)handleCall:(NSString *)payload callback:(LynxCallbackBlock)callback {
  [self.runtime handleCall:payload callback:callback];
}

@end

File: packages/runtime/android/src/main/java/org/lynxcapacitor/runtime/LynxCapacitorBridge.kt
package org.lynxcapacitor.runtime

import android.content.Context
import com.lynx.jsbridge.LynxMethod
import com.lynx.jsbridge.LynxModule
import com.lynx.jsbridge.LynxNativeModule
import com.lynx.react.bridge.Callback
import com.lynx.react.bridge.JavaOnlyArray
import com.lynx.tasm.behavior.LynxContext

@LynxNativeModule(name = "CapacitorBridge")
class LynxCapacitorBridge(context: Context) : LynxModule(context) {
    private val eventSender: (String) -> Unit = { resultJson ->
        (context as? LynxContext)?.sendGlobalEvent(
            RESULT_EVENT,
            JavaOnlyArray.of(resultJson),
        )
    }

    init {
        LynxCapacitorRuntime.setEventSender(eventSender)
    }

    @LynxMethod
    fun getPlatform(): String = "android"

    @LynxMethod
    fun getPluginHeaders(): String = LynxCapacitorRuntime.pluginHeaders()

    @LynxMethod
    fun handleCall(payload: String, callback: Callback) {
        LynxCapacitorRuntime.handleCall(payload, callback)
    }

    override fun destroy() {
        LynxCapacitorRuntime.clearEventSender(eventSender)
    }

    private companion object {
        const val RESULT_EVENT = "lynx-capacitor-result"
    }
}
